### PR TITLE
IDENTITY-6527: Unit tests for DefaultRequestPathBasedSequenceHandler

### PR DESCRIPTION
**Pull request by mefarazath**
### Proposed changes in this pull request

- Fixing the logic to get Service Provider mapped roles for local roles of a user when using RequestPath Authentication.
- Fixing possible URL encode error if sessionDataKey is null
- Unit tests

### When should this PR be merged

After Review.

### Follow up actions

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandler.java [hunk 1]
```diff
@@ -222,31 +225,50 @@ protected void handlePostAuthentication(HttpServletRequest request,
...
                 }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("No local roles to map to Service Provider role mappings. Sending back all local roles " +
+                            "as service provider mapped roles.");
+                }
+                // we don't have any sp role mappings
+                spMappedRoleList = locallyMappedUserRoles;
             }
+            spMappedRoles = StringUtils.join(spMappedRoleList.toArray(), ",");
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandler.java]: Need to check on using the Multi Attribute Separator constant here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandler.java [hunk 2]
```diff
@@ -222,31 +225,50 @@ protected void handlePostAuthentication(HttpServletRequest request,
...
-
-            boolean roleMappingDefined = false;
+        if (log.isDebugEnabled()) {
+            AuthenticatedUser authenticatedUser = sequenceConfig.getAuthenticatedUser();
+            String serviceProvider = sequenceConfig.getApplicationConfig().getApplicationName();
+            log.debug("Getting Service Provider mapped roles of application: " + serviceProvider +
+                    " of user: " + authenticatedUser);
+        }
 
+        // comma separated SP role mapped role values
```
Comment (GDRDABARERA) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandler.java]: The line comments should be in the standard way right.. 
eg. // This is a comment.

Comment (mefarazath) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandler.java]: Yes. Need to start with a Capital letter and end with a fullstop.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        Assert.assertEquals(roleClaim, FrameworkConstants.LOCAL_ROLE_CLAIM_URI);
+
+        reset(appConfig);
+        when(appConfig.getClaimMappings()).thenReturn(null);
+        // Get the default role claim URI
+        roleClaim = sequenceHandler.getSpRoleClaimUri(appConfig);
+        Assert.assertEquals(roleClaim, FrameworkConstants.LOCAL_ROLE_CLAIM_URI);
+
+    }
+
```
Comment (GDRDABARERA) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: unnecessary line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        // No IDP mapped Role claim or SP mapped Role claim available
+        reset(appConfig);
+        when(appConfig.getRoleClaim()).thenReturn(null);
+        // Get the default role claim URI
+        roleClaim = sequenceHandler.getSpRoleClaimUri(appConfig);
+        Assert.assertEquals(roleClaim, FrameworkConstants.LOCAL_ROLE_CLAIM_URI);
+    }
+
+    /*
```
Comment (GDRDABARERA) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: Use java standard method comment.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        return new Object[][]{
+                {spRoleMappings, localUserRoles, "SP_ROLE1,ADMIN,SP_ROLE2"},
+                {null, localUserRoles, localRoles},
+                {new HashMap<>(), localUserRoles, localRoles},
+                {spRoleMappings, new ArrayList<>(), null},
+                {spRoleMappings, null, null},
+                {null, null, null}
+        };
+    }
+
```
Comment (GDRDABARERA) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: Unnecessary line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+        Assert.assertNotNull(authenticatedIdPData.getUser());
+        Assert.assertEquals(authenticatedIdPData.getUser().getAuthenticatedSubjectIdentifier(), subjectIdentifier);
+        Assert.assertEquals(authenticatedIdPData.getAuthenticator(), authenticatorConfig);
+
+    }
+
+    @Test
+    public void testHandlePostAuthentication() throws Exception {
+    }
+
```
Comment (GDRDABARERA) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: Unnecessary line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java (hunk 5 above)
Comment (mefarazath) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: Done

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java (hunk 6 above)
Comment (mefarazath) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultRequestPathBasedSequenceHandlerTest.java]: done!

**Comment by mefarazath** [comment]
Fixed the issues. You can go ahead with reviewing.

**Comment by emswbandara** [comment]
Fixes [IDENTITY-6875](https://wso2.org/jira/browse/IDENTITY-6875)

**Pull request merged**